### PR TITLE
Used `google-auth` Deprecation fixed

### LINKED ISSUE #8926: Python library OAuth2Client deprecated
The [OAuth2Client](https://github.com/googleapis/oauth2client) library has been deprecated by Google since 2018, and there is a replacement library [google-auth](https://github.com/googleapis/google-auth-library-python).

https://github.com/certbot/certbot/blob/8b610239bfcf7aac06f6e36d09a5abba3ba87047/certbot-dns-google/certbot_dns_google/_internal/dns_google.py#L8

### PR DESCRIPTION
**Pull request by soham4abc**
## Pull Request Checklist
fixes #8926 

- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [x] Include your name in `AUTHORS.md` if you like.

### CODE REVIEW

**Comment by ohemorange** [comment]
Tests are failing on this PR; please let us know if you need help figuring out how to see test results, or understanding what test failures mean

**Comment by soham4abc** [comment]
> Tests are failing on this PR; please let us know if you need help figuring out how to see test results, or understanding what test failures mean
Yes that would be helpful to see where this is going wrong! I did consult many articles before making these changes but, cannot figure out where it is failing!

File: certbot-dns-google/certbot_dns_google/_internal/dns_google.py
Comment (alexzorin) [inline certbot-dns-google/certbot_dns_google/_internal/dns_google.py:85]: Looking at https://google-auth.readthedocs.io/en/master/reference/google.oauth2.service_account.html#google.oauth2.service_account.Credentials, `Credentials.from_json_keyfile_name` no longer exists.

It looks like `from_service_account_file` might be the relevant replacement. There's an example there also on how to pass the `scopes` correctly.

Once this is working, the same change should be made to the tests, to mock the new function instead of `from_json_keyfile_name`.

File: certbot-dns-google/tests/dns_google_test.py
Comment (alexzorin) [reply certbot-dns-google/tests/dns_google_test.py:110]: See also my other comment about `from_json_keyfile_name` no longer existing.

File: tools/requirements.txt
Comment (alexzorin) [inline tools/requirements.txt:95]: If you take a look at the comments at the top of this file:

https://github.com/certbot/certbot/blob/7ede5c348731558fdd2aacbbf7796c3fea4fa904/tools/requirements.txt#L1-L2

it is automatically generated by running `tools/pinning/pin.sh`.

I believe you just need to run that script rather than manually modifying this file. I am not 100% on this so I am doubling checking that advice ^^.

File: certbot/CHANGELOG.md
Comment (alexzorin) [inline certbot/CHANGELOG.md:512]: I don't know this was an accidental search-and-replace, but past CHANGELOG entries should not change.

You could add a comment to the top of the CHANGELOG file indicating that `certbot-dns-google` now depends on `google-auth` rather than `oauth2client`.

**Review by alexzorin — CHANGES_REQUESTED** [review]
Hi @soham4abc,

Thanks for working on this. I've left some comments that should help you work through the test failures and other issues.

File: tools/requirements.txt
Comment (soham4abc) [reply tools/requirements.txt]: I ran but there was no changes in the file

**Comment by soham4abc** [comment]
Only one test case failed with error 

`ERROR: InvocationError for command /home/vsts/work/1/s/.tox/lint/bin/python -m pylint --reports=n --rcfile=.pylintrc acme/acme certbot/certbot certbot-ci/certbot_integration_tests certbot-apache/certbot_apache certbot-compatibility-test/certbot_compatibility_test certbot-dns-cloudflare/certbot_dns_cloudflare certbot-dns-cloudxns/certbot_dns_cloudxns certbot-dns-digitalocean/certbot_dns_digitalocean certbot-dns-dnsimple/certbot_dns_dnsimple certbot-dns-dnsmadeeasy/certbot_dns_dnsmadeeasy certbot-dns-gehirn/certbot_dns_gehirn certbot-dns-google/certbot_dns_google certbot-dns-linode/certbot_dns_linode certbot-dns-luadns/certbot_dns_luadns certbot-dns-nsone/certbot_dns_nsone certbot-dns-ovh/certbot_dns_ovh certbot-dns-rfc2136/certbot_dns_rfc2136 certbot-dns-route53/certbot_dns_route53 certbot-dns-sakuracloud/certbot_dns_sakuracloud certbot-nginx/certbot_nginx tests/lock_test.py (exited with code 2)`

@alexzorin  Please guide!

File: certbot-dns-google/certbot_dns_google/_internal/dns_google.py
Comment (alexzorin) [inline certbot-dns-google/certbot_dns_google/_internal/dns_google.py:85]: So, the error here from `tox -e lint` is:

    certbot-dns-google/certbot_dns_google/_internal/dns_google.py:85:30: E1121: Too many positional arguments for classmethod call (too-many-function-args)

Looking at https://google-auth.readthedocs.io/en/master/reference/google.oauth2.service_account.html#google.oauth2.service_account.Credentials, `scopes` is now a keyword argument, rather than a positional one. So this line should be:

```suggestion
                credentials = Credentials.from_service_account_file(account_json, scopes=scopes)
```

File: certbot/CHANGELOG.md
Comment (alexzorin) [inline certbot/CHANGELOG.md:13]: Certbot has had a new release (1.18) since you opened your PR.

Since the CHANGELOG file is changed, you will need to pull the `master` branch and merge it into your PR. Or rebase your PR on the updated `master` branch. Sorry! I know that's annoying.

Once you've done that, move the log entry into the 1.18 section.

Please also add a space between `*` and the change text.

File: tools/requirements.txt
Comment (alexzorin) [inline tools/requirements.txt:94]: Are you sure that running `tools/pinning/pin.sh` does not update this file?

What is the output of the command? Are you running it with the Python virtualenv activated?

**Review by alexzorin — CHANGES_REQUESTED** [review]
Thanks! This is looking close to complete. Just a couple more changes.

File: tools/requirements.txt
Comment (soham4abc) [reply tools/requirements.txt:94]: Yes I did it with venv activated. The git bash window appears for a brief time with no output.

Comment (soham4abc) [reply tools/requirements.txt:94]: Don't mind I made it work! Thanks for the support!

**Comment by soham4abc** [comment]
@alexzorin  Now we getting this error ` Could not find a version that satisfies the requirement cffi==1.14.6`

**Comment by soham4abc** [comment]
@alexzorin  Did the requested changes but now more tests are failing!  :''-(

**Review by alexzorin — CHANGES_REQUESTED** [review]
Glad to see you got `pin.sh` to work!

With the new test failure, it's due to the change in that `Credentials.from_service_account_file` function to pass `scopes=scopes` rather than just `scopes`.

The test assertion [here](https://github.com/certbot/certbot/pull/8928/files#diff-e76403a7a84a73efb870304b52fc935b6609800b4636b528a92d5ebd4f5c3173R136) needs the same change to also be a keyword parameter, so `scopes=mock.ANY` rather than just `mock.ANY`.

Everything else is looking good I think.

**Comment by soham4abc** [comment]
@alexzorin Still lint check is failing. Investigating the issue!

**Comment by alexzorin** [comment]
re: the pylint errors, they're basically unrelated to your changes but are caused by pylint getting bumped to a new version. I think we should probably fix these in `master` before merging this PR. I'll try come back to it soon.

**Comment by soham4abc** [comment]
> re: the pylint errors, they're basically unrelated to your changes but are caused by pylint getting bumped to a new version. I think we should probably fix these in `master` before merging this PR. I'll try come back to it soon.

Thank you! do let me know if any changes are required!

**Comment by soham4abc** [comment]
Is the PyLint errors still stuck @alexzorin

**Comment by alexzorin** [comment]
Whenever the above pylint PR is resolved, we'll also be able to merge this PR.

**Comment by alexzorin** [comment]
Hey the pylint stuff got fixed, so I think all that is left is to resolve the resulting merge conflicts, by merging the updated `master` branch:

- Regenerate `tools/requirements.txt` (you also shouldn't get that `cffi` error anymore)
- Merging the `certbot/CHANGELOG.md` entries

**Comment by soham4abc** [comment]
@alexzorin  All updated, Please do have a look!

**Comment by soham4abc** [comment]
@alexzorin  any more changes required?

**Comment by alexzorin** [comment]
Hey, I haven't forgotten this, there's just some extra complications around dependencies that I'm looking into.

**Comment by alexzorin** [comment]
@ohemorange as I hinted at today, there are a couple of packages (`python3-googleapi` and `python3-uritemplate`) that will cause problems if our aim is packagability on Bionic and Buster. Plus one more (`python3-httplib2`) which is installable in CentOS 8 only if you enable the PowerTools repo first.

Based on reading its changelog, we need `python3-google-api-client>=1.6.1` because that's when they added support for plugging in `google-auth` over `oauth2client`. However, that is only available in `buster-backports` and not in Bionic at all. 

`python3-uritemplate` is required by `python3-google-api-client` and is only in Bullseye and Hirsute.

With PowerTools, it's not required *so far*  to install Certbot from EPEL 8, right? Not a great feeling to add that extra friction here.

| Package                                      | Required          | Buster    | Bionic    | EPEL 8         | Oldest  | 
|----------------------------------------------|-------------------|-----------|-----------|----------------|---------| 
| python3-google-auth                          | Any?                 | 1.5.1-1   | 1.3.0-2   | 1.1.1          | 1.1.1   | 
|                                              |                   |           |           |                |         | 
| python3-pyasn1                               | >=0.1.7           | 0.4.2-3   | 0.4.2-3   | 0.3.7          | 0.3.7   | 
| python3-pyasn1-modules                       | >=0.0.5           | 0.2.1-0.2 | 0.2.1-0.2 | 0.3.7          | 0.3.7   | 
| python3-rsa                                  | >=3.1.4           | 4.0-2     | 3.4.2-1   | 3.4.2          | 3.4.2   | 
| python3-six                                  | >=1.9.0           | 1.12.0-1  | 1.11.0-2  | 1.11.0         | 1.11.0  | 
| python3-cachetools                           | >=2.0.0           | 3.1.0-2   | 2.0.0-2   | 3.1.1          | 2.0.0-2 | 
|                                              |                   |           |           |                |         | 
| python3-google-api-client/python3-googleapi | >=1.6.1           | **1.5.5-1***  | **1.5.5-1**   | 1.6.7          | **?**       | 
| python3-httplib2                             | >=0.9.2,<1dev     | 0.11.3-2  | 0.9.2     | **?** (Powertools) | **?**       | 
| python3-oauth2client                         | >=1.5.0,<5.0.0dev | 4.1.2-3   | 4.1.2-3   | 4.1.3          | 4.1.2-3 | 
| python3-uritemplate                          | >=3.0.0,<4dev     | **0.6-3**     | **0.6-1**     | 3.0.0          | **?**       |

**Comment by ohemorange** [comment]
Ok, you might have seen the discussion in mattermost. To summarize, we're only going to constrain by EPEL 8 going forward. As for PowerTools -- I'm not familiar with the patterns around that. Is it likely that the need to install PowerTools will be gone in the future when the distros catch up, or will that dependency always only live there, and we'll have to add it if we ever want it?

**Comment by alexzorin** [comment]
>or will that dependency always only live there, and we'll have to add it if we ever want it?

I think it's this.

PowerTools is called "CodeReady Linux Builder" in RHEL8 (https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/package_manifest/codereadylinuxbuilder-repository) and that's where they've made `python3-httplib2` available.

Previously, `python3-httplib2` was just available in EPEL7.

I couldn't really find an explanation for why the separation was introduced. Maybe since Red Hat decided to package it themselves, EPEL decided not to?

**Comment by alexzorin** [comment]
So at the moment I am working in this branch: https://github.com/certbot/certbot/commits/test-google-constraints to figure out what things should look like with the new pinning system and strategy. 

Once that is all passing, we should be able to slap it on top of this PR and be good to go.

**Comment by alexzorin** [comment]
It looks like we also need [`google-auth-httplib2`](https://github.com/googleapis/google-auth-library-python-httplib2), if want to combine `google-api-python-client`  + `google-auth`. 

I caught this very late during manual testing, because it's a runtime dependency in `google-api-python-client==1.6.7`. It became an explicit dependency in `1.7.0`, which is not available in EPEL8.

If we still wanna go ahead with this change, we'd need EPEL to package `google-auth-httplib2`.

**Comment by ohemorange** [comment]
Oh no, nice catch. How was one packaged in EPEL without the other?

**Comment by alexzorin** [comment]
>How was one packaged in EPEL without the other?

My understanding ...

`google-auth` is useful independently of `google-api-python-client`, like, [you can just use it as an authenticator adapter for `requests`](https://google-auth.readthedocs.io/en/master/user-guide.html#requests).

`google-api-python-client` requires `httplib2`, so if you want to use `google-auth` with it, you also need the `http2` adapter for `google-auth`.

Another tidbit is that Google now recommends using https://googleapis.dev/python/dns/latest/index.html instead of the monolithic `google-api-python-client` ... but of course that's not packaged either.

**Comment by soham4abc** [comment]
Any changes required?

**Comment by alexzorin** [comment]
Hi @soham4abc,

We need to put this one on hold because making this change would create some trouble for our distro packagers. Although `oauth2client` is deprecated, it's going to keep working for the foreseeable future and we're going to stick with it for a little while longer.

Sorry! I appreciate you put a lot of work into this.

**Comment by soham4abc** [comment]
> Hi @soham4abc,
> 
> We need to put this one on hold because making this change would create some trouble for our distro packagers. Although `oauth2client` is deprecated, it's going to keep working for the foreseeable future and we're going to stick with it for a little while longer.
> 
> Sorry! I appreciate you put a lot of work into this.

No problem at all! Take your time!

**Comment by bmw** [comment]
I think we may be unblocked here since EPEL 8 uses Python 3.6 which we no longer support. Does that seem right to people?

**Comment by alexzorin** [comment]
If we are still supporting Python 3.7, that includes Debian Buster. The version of https://packages.debian.org/buster-backports/python3-googleapi that we need is only available from `buster-backports`, not `buster`. 

Is that a problem?  If backports is not a default repo, we can't require that version, right?

**Comment by bmw** [comment]
I agree with you Alex and this is unblocked. Thanks for taking a look.

(As for Debian Buster, I think one can make an argument either way, but the last time we talked about this kind of thing in general was at https://github.com/certbot/certbot/pull/8999#discussion_r704852799 where I think we largely agreed to not worry about our packages that are not seeing regularly updates. Plus, [Python 3.7 reaches its EOL next year](https://devguide.python.org/versions/) so I doubt we'll want to support them much longer anyway).

**Comment by wgreenberg** [comment]
just checking the status of this: it looks like we're not worried about the package dependency issues, and that @alexzorin's last review comments were addressed -- does that seem right, and is this ready to merge?

**Comment by alexzorin** [comment]
@wgreenberg it's close - I tried merging `master` into this PR and it seems to work okay. There are weird `mypy` failures which I don't understand, since the version of `mypy` didn't get bumped in the repin.

I do think this still needs a thorough manual test, including via instance authentication on a GCP virtual machine. I have only tested this very lightly.

The PR with merged `master` can be found at https://github.com/certbot/certbot/compare/master...alexzorin:updated_pr_8928.

If you feel like picking this up, I think it should be ready to go.

**Comment by soham4abc** [comment]
This PR seems to have a lot of rebase issues. If wanted I can issue a fresh PR with all the required changes

**Pull request closed without merging**
